src/datastructures.jl

```diff
@@ -285,3 +285,534 @@ as `TimeProfile` or in operational period `t`.
+  (typically million cubic meters).
+- **`vol_inflow::TimeProfile`** is the water inflow to the reservoir
+  (typically million cubic per time unit).
+- **`stor_res::ResourceCarrier`** is the stored `Resource`.
+- **`data::Vector{<:Data}`** is the additional data (*e.g.*, for investments or constraints
+  through [`AbstractConstraintType`](@ref)). The field `data` is conditional through usage
+  of a constructor.
+"""
+struct HydroReservoir{T} <: EMB.Storage{T}
+    id::Any
+    vol::EMB.UnionCapacity
+    vol_inflow::TimeProfile
+    stor_res::ResourceCarrier # Water
+    input::Dict{<:Resource,<:Real} # Water
+    output::Dict{<:Resource,<:Real} # Water
+    data::Vector{<:Data}
+end
+function HydroReservoir{T}(
+    id::Any,
+    vol::EMB.UnionCapacity,
+    vol_inflow::TimeProfile,
+    stor_res::ResourceCarrier,
+    data::Vector{<:Data}
+) where {T<:EMB.StorageBehavior}
+    HydroReservoir{T}(
+        id,
+        vol,
+        vol_inflow,
+        stor_res,
+        Dict(stor_res => 1.0),
+        Dict(stor_res => 1.0),
+        data
+    )
+end
+function HydroReservoir{T}(
+    id::Any,
+    vol::EMB.UnionCapacity,
+    vol_inflow::TimeProfile,
+    stor_res::ResourceCarrier
+    ) where {T<:EMB.StorageBehavior}
+    return HydroReservoir{T}(
+        id,
+        vol,
+        vol_inflow,
+        stor_res,
+        Data[],
+    )
+end
+
+"""
+    level(n::HydroReservoir)
+    level(n::HydroReservoir, t)
+
+Returns the `vol` parameter field of the HydroReservoir `n` either as `TimeProfile` or in
+operational period `t`.
+"""
+EMB.level(n::HydroReservoir) = n.vol
+EMB.level(n::HydroReservoir, t) = n.vol[t]
+
+"""
+    vol_inflow(n::HydroReservoir)
+    vol_inflow(n::HydroReservoir, t)
+
+Returns the inflow to a HydroReservoir `n` either as `TimeProfile` or in operational period
+`t`.
+"""
+vol_inflow(n::HydroReservoir) = n.vol_inflow
+vol_inflow(n::HydroReservoir, t) = n.vol_inflow[t]
+
+"""
+    HydroGate <: EMB.NetworkNode
+
+A hydro gate, modelled as a `NetworkNode` node.
+
+It an be used to model outlets/inlets and minimum/maximum requirements for water flow
+between individual reservoirs without power generation..
+
+## Fields
+- **`id`** is the name/identifier of the node.
+- **`cap::TimeProfile`** is the installed discharge capacity.
+- **`opex_var::TimeProfile`** is the variational operational costs per water flow through
+  the gate.
+- **`opex_fixed::TimeProfile`** is the fixed operational costs.
+- **`resource<:Resource`** is the water resource type since gates are only used for
+  dispatching water.
+- **`data::Vector{<:Data}`** is the additional data (*e.g.*, for investments or constraints
+  through [`AbstractConstraintType`](@ref)). The field `data` is conditional through usage
+  of a constructor.
+"""
+struct HydroGate <: EMB.NetworkNode
+    id::Any
+    cap::TimeProfile
+    opex_var::TimeProfile
+    opex_fixed::TimeProfile
+    input::Dict{<:Resource,<:Real}
```

Comment:
We should in this case remove the fields directly and instead dispatch on the functions `EMB.inputs` and `EMB.outputs`.
This is the case for all new nodes.

Reply:
I will implement this.

Reply:
Included in https://github.com/EnergyModelsX/EnergyModelsRenewableProducers.jl/pull/35/commits/2d39ebfd9ea718cc89ee4de166bd27d565647d5c